### PR TITLE
Add dictionaryOf method

### PR DESCRIPTION
**Pull request by glennrfisher**
### Overview

This pull request adds a `dictionaryOf(_:type:)` function that is analogous to the existing `arrayOf(_:type:)` function.

### Example

This pull request makes the following example possible:

```swift
let residentJSON = JSON.Dictionary([
    "residentsByName": [
        "Matt": ["name": "Matt", "age": 33, "hasPet": false, "rent": .Null],
        "Drew": ["name": "Drew", "hasPet": true, "rent": 1234.5],
        "Pat": ["name": "Pat", "age": 28, "hasPet": .Null]
    ]
])

do {
    let residents = try residentJSON.dictionaryOf("residentsByName", type: Resident.self)
    if let matt = residents["Matt"] {
        print("Matt is \(matt.age) years old.")
    }
} catch {
    // handle error
}
```

### Justification

I admit that it's rare to run into JSON that would make use of the `dictionaryOf(_:type:)` function. However, I needed something similar to decode [this parameter](https://github.com/watson-developer-cloud/ios-sdk/blob/master/Source/SpeechToTextV1/Models/TranscriptionResult.swift#L34) in the [IBM Watson Developer Cloud iOS SDK](https://github.com/watson-developer-cloud/ios-sdk).

The [IBM Watson Speech to Text documentation](https://www.ibm.com/smarterplanet/us/en/ibmwatson/developercloud/speech-to-text/api/v1/#recognize_audio_sessions18) provides an example JSON payload for which the `dictionaryOf(_:type:)` function would be very helpful:

```swift
{
  "keywords_result": {
     "tornadoes": [
        {
           "normalized_text": "tornadoes",
           "start_time": 1.52,
           "confidence": 1.0,
           "end_time": 2.15
        }
     ],
     "colorado": [
        {
           "normalized_text": "Colorado",
           "start_time": 4.94,
           "confidence": 0.925,
           "end_time": 5.62
        }
     ]
  }
}
```

In particular, I believe the following should be able to decode the JSON payload:

```swift
public struct TranscriptionResult: JSONDecodable {

    public let keywordResults: [String: [KeywordResult]]

    public init(json: JSON) throws {
        keywordResults = try json.dictionaryOf("keywords_result")
    }
}
```

### Implementation Notes

I tried to follow the example of the `arrayOf(_:type:)` function. I used it (and its documentation) as the starting point for `dictionaryOf(_:type:)`. Hopefully, these changes are consistent with your style as a result.

I also searched for all occurrences of `arrayOf` so I could add an analogous use of `dictionaryOf`.

### CODE REVIEW

File: Sources/JSONDecodable.swift [hunk 1]
```diff
@@ -148,4 +148,24 @@ internal extension JSON {
...
+    ///                   to disambiguate the value type to decode with.
+    /// - returns: A `Dictionary` of string keys and decoded values.
+    /// - throws: One of the `JSON.Error` cases thrown by `decode(_:type:)` or
+    ///           any error that arises from decoding the contained values.
+    /// - seealso: `JSON.decode(_:type:)`
+    static func getDictionaryOf<Decoded: JSONDecodable>(json: JSON) throws -> [Swift.String: Decoded] {
+        guard case let .Dictionary(dictionary) = json else {
+            throw Error.ValueNotConvertible(value: json, to: Swift.Dictionary<Swift.String, JSON>)
+        }
+        var decodedDictionary: [Swift.String: Decoded] = [:]
```
Comment (zwaldowski) [inline Sources/JSONDecodable.swift]: Consider [`Dictionary.init(minimumCapacity:)`](https://developer.apple.com/library/ios/documentation/Swift/Reference/Swift_Dictionary_Structure/index.html).

**Comment by zwaldowski** [comment]
I'm a big 👍 on this, and will merge tomorrow save for that one comment. We appreciate the excellent tests!

File: Sources/JSONDecodable.swift [hunk 2]
```diff
@@ -148,4 +148,24 @@ internal extension JSON {
         return try getArray(json).map(Decoded.init)
     }
     
+    /// Attempts to decode many values from a descendant JSON object at a path
+    /// into JSON.
+    /// - parameter type: If the context this method is called from does not
+    ///                   make the return type clear, pass a type implementing `JSONDecodable`
+    ///                   to disambiguate the value type to decode with.
+    /// - returns: A `Dictionary` of string keys and decoded values.
```
Comment (mdmathias) [inline Sources/JSONDecodable.swift]: s/decoded/`Decoded`

File: Sources/JSONDecodable.swift [hunk 3]
```diff
@@ -148,4 +148,24 @@ internal extension JSON {
         return try getArray(json).map(Decoded.init)
     }
     
+    /// Attempts to decode many values from a descendant JSON object at a path
+    /// into JSON.
+    /// - parameter type: If the context this method is called from does not
```
Comment (mdmathias) [inline Sources/JSONDecodable.swift]: There is no `type` parameter.

Comment (mdmathias) [inline Sources/JSONDecodable.swift]: Seems like the documentation in `JSONDecodable.swift` needs a little cleaning up!  Sorry if that disorganization was confusing.

**Comment by mdmathias** [comment]
Nice work!  Thanks also for exposing some documentation problems.  : |

File: Sources/JSONDecodable.swift (hunk 3 above)
Comment (glennrfisher) [inline Sources/JSONDecodable.swift]: Good catch! Yeah, I tried to follow the documentation of the `getArrayOf` function, which includes a `type` parameter for some reason. I'll remove the `type` parameter from the `getDictionaryOf` function's documentation.

Would you also like me to remove the `type` parameter from the `getArrayOf` function's documentation or would you prefer to handle that outside of this PR?

File: Sources/JSONDecodable.swift (hunk 2 above)
Comment (glennrfisher) [inline Sources/JSONDecodable.swift]: I assume that means to substitute "decoded" for "`Decoded`"?

Again, would you like me to update the documentation for the `getArrayOf` method as well?

File: Sources/JSONDecodable.swift (hunk 1 above)
Comment (glennrfisher) [inline Sources/JSONDecodable.swift]: Great idea!

File: Sources/JSONSubscripting.swift [hunk 4]
```diff
@@ -398,6 +412,26 @@ extension JSON {
     public func dictionary(path: JSONPathType..., alongPath options: SubscriptingOptions) throws -> [Swift.String: JSON]? {
         return try mapOptionalAtPath(path, alongPath: options, transform: JSON.getDictionary)
     }
+    
+    /// Optionally attempts to decode many values from a descendant object at a path
+    /// into JSON.
+    /// - parameter path: 0 or more `String` or `Int` that subscript the `JSON`
+    /// - parameter alongPath: Options that control what should be done with values that are `null` or keys that are missing.
+    /// - returns: A `Dictionary` of `String` mapping to decoded elements if a
```
Comment (glennrfisher) [inline Sources/JSONSubscripting.swift]: Got it. Would you also like me to add a `type` parameter to the `arrayOf` function? (i.e. at line 382)

File: Sources/JSONDecodable.swift [hunk 5]
```diff
@@ -148,4 +148,24 @@ internal extension JSON {
...
+    ///                   make the return type clear, pass a type implementing `JSONDecodable`
+    ///                   to disambiguate the value type to decode with.
+    /// - returns: A `Dictionary` of string keys and decoded values.
+    /// - throws: One of the `JSON.Error` cases thrown by `decode(_:type:)` or
+    ///           any error that arises from decoding the contained values.
+    /// - seealso: `JSON.decode(_:type:)`
+    static func getDictionaryOf<Decoded: JSONDecodable>(json: JSON) throws -> [Swift.String: Decoded] {
+        guard case let .Dictionary(dictionary) = json else {
+            throw Error.ValueNotConvertible(value: json, to: Swift.Dictionary<Swift.String, JSON>)
```
Comment (glennrfisher) [inline Sources/JSONDecodable.swift]: I think the `to:` parameter should instead be `to: Swift.Dictionary<Swift.String, Decoded>`. I will update that in the PR.

**Comment by glennrfisher** [comment]
Thanks, @mdmathias!

I committed an update to reflect the changes from your comments. Thanks for the thorough review! :)

File: Sources/JSONDecodable.swift (hunk 3 above)
Comment (mdmathias) [inline Sources/JSONDecodable.swift]: I believe I addressed this in a separate PR.

File: Sources/JSONDecodable.swift (hunk 2 above)
Comment (mdmathias) [inline Sources/JSONDecodable.swift]: Other way around. Replace 'decoded' with `Decoded`.

Comment (mdmathias) [inline Sources/JSONDecodable.swift]: You can fix this one. :)

File: Sources/JSONSubscripting.swift (hunk 4 above)
Comment (mdmathias) [inline Sources/JSONSubscripting.swift]: Ya!

File: Sources/JSONDecodable.swift (hunk 5 above)
Comment (mdmathias) [inline Sources/JSONDecodable.swift]: :+1:

**Comment by glennrfisher** [comment]
Okay, I added those fixes to the documentation for `getArrayOf` and `arrayOf`. I think this is ready to be merged after the Travis build completes.

**Pull request merged**